src/measure_theory/integral/integral_eq_improper.lean

```diff
@@ -651,73 +651,78 @@ end measure_theory
-    { push_neg at u,
-      rw [Ioc_eq_empty_of_le u.le, integral_empty, le_sub, sub_zero],
-      apply f_incr,
-      { rw mem_Icc, split, tauto, exact h_lt.le, },
-      { rw mem_Icc, split, exact h_lt.le, tauto, },
-      exact h_lt.le, },
+    rcases lt_or_le (B n) (A n) with u|u,
+    { rw [Ioc_eq_empty_of_le u.le, integral_empty, sub_nonneg],
+      exact f_incr ⟨le_refl _, hab⟩ ⟨hab, le_refl _⟩ hab },
     -- now main case
     have : ∫ (x : ℝ) in Ioc (A n) (B n), ∥f' x∥ = ∫ (x : ℝ) in Ioc (A n) (B n), f' x,
-    { apply set_integral_congr, { exact measurable_set_Ioc },
-      intros x hx, dsimp only,
-      exact real.norm_of_nonneg (hpos _ $ subset_trans Ioc_subset_Icc_self (Icc_sub n) hx),},
+    { refine set_integral_congr measurable_set_Ioc (λ x hx, _),
+      exact real.norm_of_nonneg (hnonneg _ $ subset_trans Ioc_subset_Icc_self (Icc_sub n) hx),},
     rw this,
     rw [←interval_integral.integral_of_le u, interval_integral.integral_eq_sub_of_has_deriv_at],
-    swap,
-    { intros x hx, rw interval_of_le u at hx, apply hderiv x, apply Icc_sub n, exact hx, },
     { have t1a : a ≤ A n, { simp only [le_add_iff_nonneg_right], exact (t0 n).le, },
       have t1b : B n ≤ b, { apply sub_le_self, exact (t0 n).le, },
-      have t3 : f (B n) ≤ f b,
-      { apply f_incr, exact ⟨le_trans t1a u, t1b⟩, rw mem_Icc, simpa using h_lt.le, exact t1b },
-      have t4 : f a ≤ f (A n),
-      { apply f_incr, rw mem_Icc, simpa using h_lt.le, exact ⟨t1a, le_trans u t1b⟩, exact t1a, },
-      exact sub_le_sub t3 t4, },
+      exact sub_le_sub (f_incr ⟨le_trans t1a u, t1b⟩ ⟨hab, le_refl _⟩ t1b)
+         (f_incr ⟨le_refl _, hab⟩ ⟨t1a, le_trans u t1b⟩ t1a), },
+    { intros x hx, rw interval_of_le u at hx, apply hderiv x, apply Icc_sub n, exact hx, },
     { refine (continuous_on.mono hcontf' _).interval_integrable,
       rw interval_of_le u, exact Icc_sub n, }, }
   end
 
-/-- Fundamental theorem of calculus-2: If `f : ℝ → ℝ` has a derivative `f' x` for all `x` in
-  `(a, b)` and is continuous on `[a, b]`, and `f'` is continuous and non-negative on `(a, b)`, then
-  `∫ y in a..b, f' y = f(b) - f(a)`. -/
 theorem interval_integral.integral_eq_sub_of_has_deriv_at_of_nonneg {a b : ℝ} {f f' : ℝ → ℝ}
-  (hderiv : ∀ x ∈ Ioo a b, has_deriv_at f (f' x) x) (hpos : ∀ x ∈ Ioo a b, 0 ≤ f' x)
-  (hcontf : continuous_on f $ Icc a b) (hcontf' : continuous_on f' $ Ioo a b) (hab : a ≤ b) :
-  ∫ y in a..b, f' y = f b - f a :=
-interval_integral.integral_eq_sub_of_has_deriv_at_of_le hab hcontf hderiv (
-  (interval_integrable_iff_integrable_Ioc_of_le hab).mpr
-  (interval_integral.integrable_of_has_deriv_at_of_nonneg hderiv hpos hcontf hcontf'))
+  (hderiv : ∀ x ∈ interior (interval a b), has_deriv_at f (f' x) x)
+  (hnonneg : ∀ x ∈ interior (interval a b), 0 ≤ f' x)
+  (hcontf : continuous_on f $ interval a b)
+  (hcontf' : continuous_on f' $ interior (interval a b)) :
+  (interval_integrable f' volume a b) ∧ (∫ y in a..b, f' y = f b - f a) :=
```

Comment:
I don't know what the preferred style of the library is: Is it better to return this conjunction or is it better to return the conjunction as a private lemma and then have two theorems: one that returns the integrability and one that returns the value?